gptme/llm/llm_anthropic.py

```diff
@@ -286,3 +282,81 @@ def _spec2tool(
         "description": spec.get_instructions("tool"),
         "input_schema": parameters2dict(spec.parameters),
     }
+
+
+def _prepare_messages_for_api(
+    messages: list[Message], tools: list[ToolSpec] | None
+) -> tuple[
+    list["anthropic.types.beta.prompt_caching.PromptCachingBetaMessageParam"],
+    list["anthropic.types.beta.prompt_caching.PromptCachingBetaTextBlockParam"],
+    list["anthropic.types.beta.prompt_caching.PromptCachingBetaToolParam"] | None,
+]:
+    """Prepare messages for the Anthropic API.
+
+    This function:
+    1. Transforms system messages
+    2. Handles file attachments
+    3. Applies cache control
+    4. Prepares tools
+
+    Args:
+        messages: List of messages to prepare
+        tools: List of tool specifications
+
+    Returns:
+        tuple containing:
+        - Prepared message dictionaries
+        - System messages
+        - Tool dictionaries (if tools provided)
+    """
+    # noreorder
+    from anthropic.types.beta.prompt_caching import (  # fmt: skip
+        PromptCachingBetaImageBlockParam,
+        PromptCachingBetaMessageParam,
+        PromptCachingBetaTextBlockParam,
+        PromptCachingBetaToolResultBlockParam,
+        PromptCachingBetaToolUseBlockParam,
+    )
+
+    # Transform system messages
+    messages, system_messages = _transform_system_messages(messages)
+
+    # Handle files and convert to dicts
+    messages_dicts = _handle_files(msgs2dicts(messages))
+
+    # Apply cache control to optimize performance
+    messages_dicts_new: list[PromptCachingBetaMessageParam] = []
+    for i, msg in enumerate(messages_dicts):
+        content_parts: list[
+            PromptCachingBetaTextBlockParam
+            | PromptCachingBetaImageBlockParam
+            | PromptCachingBetaToolUseBlockParam
+            | PromptCachingBetaToolResultBlockParam
+        ] = []
+        raw_content = (
+            msg["content"]
+            if isinstance(msg["content"], list)
+            else [{"type": "text", "text": msg["content"]}]
+        )
+
+        for part in raw_content:
+            if isinstance(part, dict):
+                if part.get("type") == "text" and i == len(messages_dicts) - 1:
```

Comment:
Might want to review this later to cache at the optimal locations, especially considering RAG and "live" messages. Should probably also add a cache step at the system prompt.